transaction.go

```diff
@@ -55,28 +82,149 @@ func (tx *readOnlyTransaction) Rollback() error {
+// that have been executed on a read/write transaction. These statements must
+// implement a retry method that will be executed during a transaction retry.
+type retriableStatement interface {
+	// retry retries the statement on a new Spanner transaction. The method must
+	// return nil if it receives the same result as during the initial attempt,
+	// and otherwise ErrAbortedDueToConcurrentModification.
+	//
+	// Note: This method does not return any error that is returned by Spanner
+	// when the statement is executed. Instead, if the statement returns an
+	// error, the returned error should be compared with the result during the
+	// initial attempt. If the two errors are equal, the retry of the statement
+	// should be considered successful and the method should return nil.
+	retry(ctx context.Context, tx *spanner.ReadWriteStmtBasedTransaction) error
+}
+
+// retriableUpdate implements retriableStatement for update statements.
+type retriableUpdate struct {
+	// stmt is the statement that was executed on Spanner.
+	stmt spanner.Statement
+	// c is the record count that was returned by Spanner.
+	c int64
+	// err is the error that was returned by Spanner.
+	err error
 }
 
+// retry retries an update statement on Spanner. It returns nil if the result
+// of the statement during the retry is equal to the result during the initial
+// attempt.
+func (ru *retriableUpdate) retry(ctx context.Context, tx *spanner.ReadWriteStmtBasedTransaction) error {
+	c, err := tx.Update(ctx, ru.stmt)
+	if err != nil && spanner.ErrCode(err) == codes.Aborted {
+		return err
+	}
+	if !errorsEqualForRetry(err, ru.err) {
+		return ErrAbortedDueToConcurrentModification
+	}
+	if c != ru.c {
+		return ErrAbortedDueToConcurrentModification
+	}
+	return nil
+}
+
+// runWithRetry executes a statement on a go/sql read/write transaction and
+// automatically retries the entire transaction if the statement returns an
+// Aborted error. The method will return ErrAbortedDueToConcurrentModification
+// if the transaction is aborted and the retry fails because the retry attempt
+// returned different results than the initial attempt.
+func (tx *readWriteTransaction) runWithRetry(ctx context.Context, f func(ctx context.Context) error) (err error) {
+	for {
+		if err == nil {
+			err = f(ctx)
+		}
+		if err == ErrAbortedDueToConcurrentModification {
+			return
+		}
+		if spanner.ErrCode(err) == codes.Aborted {
+			err = tx.retry(ctx)
+			continue
+		}
+		return
+	}
+}
+
+// retry retries the entire read/write transaction on a new Spanner transaction.
+// It will return ErrAbortedDueToConcurrentModification if the retry fails.
+func (tx *readWriteTransaction) retry(ctx context.Context) (err error) {
+	tx.rwTx, err = spanner.NewReadWriteStmtBasedTransaction(ctx, tx.client)
+	if err != nil {
+		return err
+	}
+	for _, stmt := range tx.statements {
+		err = stmt.retry(ctx, tx.rwTx)
+		if err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+// Commit implements driver.Tx#Commit().
```

Comment:
Should we move the second line in the comment to the first line? 

```go
// Commit implements driver.Tx#Commit(). It will commit the underlying ...
...
```

Reply:
The godoc convention that the first sentence should start with the name of the method, so that would not work with that.